runtime-transaction/src/compute_budget_instruction_details.rs

```diff
@@ -15,39 +17,90 @@ use {
 
 #[cfg_attr(test, derive(Eq, PartialEq))]
 #[cfg_attr(feature = "dev-context-only-utils", derive(Clone))]
-#[derive(Default, Debug)]
+#[derive(Debug)]
 pub(crate) struct ComputeBudgetInstructionDetails {
     // compute-budget instruction details:
     // the first field in tuple is instruction index, second field is the unsanitized value set by user
     requested_compute_unit_limit: Option<(u8, u32)>,
     requested_compute_unit_price: Option<(u8, u64)>,
     requested_heap_size: Option<(u8, u32)>,
     requested_loaded_accounts_data_size_limit: Option<(u8, u32)>,
-    num_non_compute_budget_instructions: u32,
+    // instructions counts
+    num_compute_budget_instructions: u32,
+    num_builtin_instructions: u32,
+    num_non_builtin_instructions: u32,
+    // The vector of counters, matching the size of the static vector MIGRATION_FEATURE_IDS,
+    // tracks the usage frequency of each feature ID, with each counter representing the
+    // number of times its corresponding feature ID is referenced in this transaction.
+    migrating_builtin: Vec<u32>,
+}
+
+impl Default for ComputeBudgetInstructionDetails {
+    fn default() -> Self {
+        Self {
+            requested_compute_unit_limit: None,
+            requested_compute_unit_price: None,
+            requested_heap_size: None,
+            requested_loaded_accounts_data_size_limit: None,
+            num_compute_budget_instructions: 0,
+            num_builtin_instructions: 0,
+            num_non_builtin_instructions: 0,
+            migrating_builtin: vec![0; MIGRATION_FEATURES_ID.len()],
+        }
+    }
 }
 
 impl ComputeBudgetInstructionDetails {
     pub fn try_from<'a>(
         instructions: impl Iterator<Item = (&'a Pubkey, SVMInstruction<'a>)>,
     ) -> Result<Self> {
         let mut filter = ComputeBudgetProgramIdFilter::new();
-
         let mut compute_budget_instruction_details = ComputeBudgetInstructionDetails::default();
         for (i, (program_id, instruction)) in instructions.enumerate() {
-            if filter.is_compute_budget_program(instruction.program_id_index as usize, program_id) {
-                compute_budget_instruction_details.process_instruction(i as u8, &instruction)?;
-            } else {
-                saturating_add_assign!(
-                    compute_budget_instruction_details.num_non_compute_budget_instructions,
-                    1
-                );
+            match filter.get_program_kind(instruction.program_id_index as usize, program_id) {
```

Comment:
If you agree with moving this to `calculate_default_compute_unit_limit`, I've implemented the pre-requisite PR (https://github.com/anza-xyz/agave/pull/3853) to unblock adding an instructions iterator param to `sanitize_and_convert_to_compute_budget_limits`

Reply:
I'm leaning towards this instead of storing a `Vec`. Re-iterating is likely less costly than allocation, and we can remove the re-iteration once the feature is activated.

Reply:
> We really don't need to do all this builtin program accounting unless the compute limit isn't specified by the transaction. 

This is great observation. Another possibility is to re-iterate in `try_from` when necessary. Something like this: 
```
struct builtin_instruction_details {
            num_builtin_instructions: 0,
            num_non_builtin_instructions: 0,
            migrating_builtin: vec![0; MIGRATION_FEATURES_ID.len()],
}

fn try_from( ixs: iter ) -> Result<Self> {
    // original iteration: 
    // to count `num_compute_budget_instructions` and process_compute_budget_instruction if present

    // possible second iter
    if requested_compute_unit_limit.is_none() {
        // iterate again to create "builtin_instruction_details"
    }
}
```

Reply:
Also, based on benchmarking, the cost of allocating vec with fixed size of 3 is much lower than the cost of hashing Pubkey to lookup at `BUILTIN_INSTRUCTION_COSTS`. Re-iterate within `try_from` will reuse cached `ComputeBudgetProgramIdFilter`, avoiding re-hashing.

Reply:
wdyt something like this: https://github.com/tao-stones/solana/commit/6dcc0cd81a037361c73e833ddf9a6539153e5bfa
If acceptable, can put out a pre-pr to add `+ Clone` to `program_instructions_iter()`.

Reply:
> wdyt something like this: https://github.com/tao-stones/solana/commit/86fcaebcf9f885b9e2a93123736692f9273a09d9 (edited)

That does fix the issue where we were unnecessarily calculating builtin counts even when the compute unit limit was specified by the transaction. But I think this implementation is too convoluted. I prefer the code to be simpler, especially if we are planning to backport this change.

> Also, based on benchmarking, the cost of allocating vec with fixed size of 3 is much lower than the cost of hashing Pubkey to lookup at BUILTIN_INSTRUCTION_COSTS. Re-iterate within try_from will reuse cached ComputeBudgetProgramIdFilter, avoiding re-hashing.

I don't think we can purely rely on benchmarks for making this decision. The extra allocation is far more more expensive while the validator is running than in an isolated benchmark.

Reply:
> That does fix the issue where we were unnecessarily calculating builtin counts even when the compute unit limit was specified by the transaction. But I think this implementation is too convoluted. I prefer the code to be simpler, especially if we are planning to backport this change.

One thing that didn't make sense about your commit is that we are still calculating `num_non_compute_budget_instructions`. Similar to the `builtin_instruction_details` field you added, it only needs to be calculated if the tx didn't specify an explicit compute unit limit.

Reply:
just a note that `num_non_compute_budget_instructions` is byproduct of first iteration when checking `is_compute_budget_program()`, so in case cu-limit is not requested and feature is not activated (eg. current behavior), it's ready to calc default cu limit by `num_non_compute_budget_instructions * 200K`.